immutablecollections/_immutableset.py

```diff
@@ -246,12 +247,28 @@ def union(
         If check top level types is provided, all elements of both sets must match the specified
         type.
         """
-        return (
-            ImmutableSet.builder(check_top_type_matches)
-            .add_all(self)
-            .add_all(other)
-            .build()
-        )
+        if check_top_type_matches:
+            for (side_name, side) in (("left", self), ("right", other)):
+                items_not_matching = [
+                    x for x in side if not isinstance(x, check_top_type_matches)
+                ]
+                if items_not_matching:
+                    raise RuntimeError(
```

Comment:
I think `TypeError` is a better choice, according to https://docs.python.org/3/library/exceptions.html#TypeError.

Reply:
Fixed in 463cbbe50706a6ee48aa83abf6aa5c6e3699715e